### PR TITLE
Error message for enum case in non-enum class companion object

### PR DESCRIPTION
**Pull request by esarbe**
see #1589

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala [hunk 1]
```diff
@@ -6,6 +6,7 @@ import core._
 import util.Positions._, Types._, Contexts._, Constants._, Names._, NameOps._, Flags._
 import SymDenotations._, Symbols._, StdNames._, Annotations._, Trees._
 import Decorators._
+import dotty.tools.dotc.reporting.diagnostic.messages.EnumCaseDefinitionInNonEnumOwner
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala]: Can we shorten this to just `reporting.diagnostic.messages.EnumCaseDefinitionInNonEnumOwner`?

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 2]
```diff
@@ -1748,7 +1748,7 @@ object messages {
     val explanation =
       hl"you have to provide either ${"class"}, ${"trait"}, ${"object"}, or ${"enum"} definitions after qualifiers"
   }
-
+
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Introduction of a trailing space?

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 3]
```diff
@@ -1770,4 +1770,21 @@ object messages {
...
   }
 
+  case class EnumCaseDefinitionInNonEnumOwner(owner: Symbol)(implicit ctx: Context)
+    extends Message(EnumCaseDefinitionInNonEnumOwnerID) {
+      val kind = "Syntax"
+      val msg = em"case not allowed here, since owner ${owner} is not an `enum' object"
+      val explanation = {
+
+        hl"""
+            | ${"case"} is only allowed at this place if the surrounding object is the companion object of an `enum class`.
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Just curious, is `${"case"}` equivalent to `case`? I see similar uses above which I don't understand.

**Comment by allanrenucci** [comment]
Before we get this in, I think we should discuss if it is a good idea or not to have links in the error messages (ref #3199).

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 2 above)
Comment (esarbe) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Uh, nice catch! Thanks.

**Comment by esarbe** [comment]
I can remove the link to circumvent the discussion. I'm quite willing to come back and re-introduce the link if this should be outcome of the discussion.

**Comment by allanrenucci** [comment]
You can make the tests pass by rebasing on top of master

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 4]
```diff
@@ -1770,4 +1770,18 @@ object messages {
...
+      val kind = "Syntax"
+      val msg = em"case not allowed here, since owner `${owner}` is not an `${"enum"}` object"
+      val explanation = {
+
+        hl"""
+            | `${"case"}` is only allowed at this place if the surrounding object is the companion object of an `${"enum"} ${"class"}`.
+            | If you wanted to create an enumeration make sure that the corresponding class has the `${"enum"}` keyword.
+            | Otherwise you might have forgotten `${"class"}` or `${"object"}` after this `${"case"}` here.
+            |
+          """.stripMargin
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: You don't need to `stripMargin` the `hl` interpolator does it for you.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 5]
```diff
@@ -1770,4 +1770,18 @@ object messages {
           |"""
   }
 
+  case class EnumCaseDefinitionInNonEnumOwner(owner: Symbol)(implicit ctx: Context)
+    extends Message(EnumCaseDefinitionInNonEnumOwnerID) {
+      val kind = "Syntax"
+      val msg = em"case not allowed here, since owner `${owner}` is not an `${"enum"}` object"
+      val explanation = {
+
+        hl"""
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Start your message here

**Comment by esarbe** [comment]
Thanks for spotting these! Should be better now.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 6]
```diff
@@ -1777,5 +1777,15 @@ object messages {
     val explanation =
       hl"""A class marked with the ${"final"} keyword cannot be extended"""
   }
-
+  case class EnumCaseDefinitionInNonEnumOwner(owner: Symbol)(implicit ctx: Context)
+    extends Message(EnumCaseDefinitionInNonEnumOwnerID) {
+      val kind = "Syntax"
+      val msg = em"case not allowed here, since owner `${owner}` is not an `${"enum"}` object"
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: You shouldn't backtick `owner` and `enum`

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 7]
```diff
@@ -1777,5 +1777,15 @@ object messages {
...
+      val explanation =
+        hl"""
+            | `${"case"}` is only allowed at this place if the surrounding object is the companion object of an `${"enum"} ${"class"}`.
+            | If you wanted to create an enumeration make sure that the corresponding class has the `${"enum"}` keyword.
+            | Otherwise you might have forgotten `${"class"}` or `${"object"}` after this `${"case"}` here.
+          """
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: I would prefer something like:
```scala
  val explanation =
    hl"""${"enum"} cases are only allowed within the companion ${"object"} of an ${"enum class"}.
        |If you want to create an ${"enum"} case, make sure the corresponding ${"enum class"} exists
        |and has the ${"enum"} keyword."""
```

Comment (esarbe) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Let's take that.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 6 above)
Comment (esarbe) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: should be fixed
